### ISSUE TITLE
Beta regression in span printing with tabs involved

### ISSUE THREAD

**Issue by est31** [rust-lang/rust#47377]
This code (containing tabs) gives an error ([playground](https://play.rust-lang.org/?gist=7fbea34512455179a57ca244b297ba26&version=nightly)):

```rust
	let b = "hello";
	let _a = b + ", World!";
```

The error works fine on stable:

```
  |
3 |  let _a = b + ", World!";
  |           ^^^^^^^^^^^^^^ `+` can't be used to concatenate two `&str` strings
help: `to_owned()` can be used to create an owned `String` from a string reference. String concatenation appends the string on the right to the string on the left and may require reallocation. This requires ownership of the string on the left
  |
3 |  let _a = b.to_owned() + ", World!";
  |           ^^^^^^^^^^^^

```

But is mispositioned on rustc 1.25.0-nightly (73ac5d6a8 2018-01-11) as well as on rustc 1.24.0-beta.2 (a19122cfa 2018-01-10):

```
  |
3 |     let _a = b + ", World!";
  |              ^^^^^^^^^^^^^^ `+` can't be used to concatenate two `&str` strings
help: `to_owned()` can be used to create an owned `String` from a string reference. String concatenation appends the string on the right to the string on the left and may require reallocation. This requires ownership of the string on the left
  |
3 |     let _a = b.to_owned() + ", World!";
  |           ^^^^^^^^^^^^

```

cc @estebank

**Comment by BubbaSheen** [comment]
Interesting. Im new to code work so following you guys has been extremely
helpful thanks

**Comment by est31** [comment]
Using bisect-rust I could see that the issue started with #45953, but probably the actual issue is somewhere else (e.g. in the machinery added in #45711 that #45953 starts using for tabs).

**Comment by est31** [comment]
Yeah, the issue also exists when emojis are involved (no tabs):
```rust
    let b = "hello";
    println!("🦀🦀🦀🦀🦀"); let _a = b + ", World!";
```
Gives:
```
4 |     println!("🦀🦀🦀🦀🦀"); let _a = b.to_owned() + ", World!";
  |                                 ^^^^^^^^^^^^
```

The emoji manifestation of the issue is no beta regression, seems to exist on stable as well.

**Comment by est31** [comment]
I've created  #47380 for the emoji/unicode issue.

**Comment by estebank** [comment]
It seems to me that the issue only manifests itself in suggestions, and that is handled separately from the main diagnostics.

---
Mentoring instructions:
-

[`emit_suggestion_default`](https://github.com/rust-lang/rust/blob/0b90e4e8cd068910f604f3e1fb5d03cc01f1658f/src/librustc_errors/emitter.rs#L1190-L1191) needs to be modified to use `col_display` instead of `col.0`, the start being `span_start_pos.col_display + start` and the end being `span_start_pos.col.0 + start + sub_len` (`sub_len` already accounts for unicode chars). This change should fix both this and #47380.

**Comment by gaurikholkar** [comment]
Will pick this up
